### PR TITLE
Add cached dna sequence extraction

### PR DESCRIPTION
**Pull request by JamesJeffryes**

### CODE REVIEW

File: lib/AssemblyAPI/AssemblySequence.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            raise ValueError(f'{loc} is not a valid location on this assembly')
+        return st
+    return 0
+
+
+def _extract_sequence(assembly_dir, locations):
+    """Given a cached assembly directory extract a sequence for each of the supplied locations"""
+    contig = None
+    fragments = []
+    for loc_tup in locations:
```
Comment (slebras) [inline lib/AssemblyAPI/AssemblySequence.py]: isnt locations at this point a tuple of <int, string, int, string>?

Comment (slebras) [inline lib/AssemblyAPI/AssemblySequence.py]: It seems to me that `locations` gets passed as a tuple of <int, string, int, string> to this function, do we want to iterate over it at this point?

File: lib/AssemblyAPI/AssemblySequence.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        loc = Location(*loc_tup)
```
Comment (slebras) [inline lib/AssemblyAPI/AssemblySequence.py:33]: never seen a nametuple used before, makes the code much more readable!

**Review by slebras — CHANGES_REQUESTED** [review]
Just one thing I'm not super sure about, but otherwise looks good

File: lib/AssemblyAPI/AssemblySequence.py (hunk 2 above)
Comment (JamesJeffryes) [reply lib/AssemblyAPI/AssemblySequence.py:33]: Now just imagine if that was the return value for WS object info! (I highly recommend using this trick for object info returns actually)

File: lib/AssemblyAPI/AssemblySequence.py (hunk 1 above)
Comment (JamesJeffryes) [reply lib/AssemblyAPI/AssemblySequence.py]: This is confusing but locations is a list of tuples that represent a single genetic feature. The function that calls this is designed to accept multiple features. The params for that fuction is the input might look like this: {'ref': self.assembly_ref_1, 'locations': [[['NZ_ALQT01000001', 10, "+", 5], ['NZ_ALQT01000001', 10, "+", 5]], [['NZ_ALQT01000001', 50, "-", 10]]]}

I'll rename to feature_locs to clarify

**Pull request merged**